gh-142466: Pixi package definitions for downstream use

lucascolley opened the pull request:
implementation for gh-142466, supporting https://github.com/numpy/numpy/pull/30381 and https://github.com/scipy/scipy/pull/24066.

lucascolley committed:
WIP: ENH: Pixi package definitions for downstream development

[skip ci]

lucascolley committed:
linux-64 support

lucascolley commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,2 @@
+- see comments for which fields should be maintained and kept up-to-date
+- combining directories blocked on https://github.com/prefix-dev/pixi/issues/4599
```
hopefully once this issue is solved, we would be able to have just one `pixi.toml` and `recipe.yaml` file, rather than separate ones in separate directories for each build variant.

lucascolley commented on .gitignore:
```diff
@@ -158,6 +158,7 @@ Tools/msi/obj
 Tools/ssl/amd64
 Tools/ssl/win32
 Tools/freeze/test/outdir
+Tools/pixi-packages/**/.pixi
```
might be easier to just ignore `.pixi` across the board

FFY00 replied:
I think that would be preferable, but I don't have a strong opinion on it.

lucascolley commented on Tools/pixi-packages/build.sh:
```diff
@@ -0,0 +1,36 @@
+make -j"${CPU_COUNT}" install
+ln -sf "${PREFIX}/bin/python${MINOR_VERSION}" "${PREFIX}/bin/python"
+
+# https://github.com/prefix-dev/rattler-build/issues/2012
+if [[ ${OSTYPE} == "darwin"* ]]; then
+    cp "${BUILD_PREFIX}/lib/clang/21/lib/darwin/libclang_rt.asan_osx_dynamic.dylib" "${PREFIX}/lib/libclang_rt.asan_osx_dynamic.dylib"
+fi
```
this should be removed once the linked issue is solved

lucascolley commented on Tools/pixi-packages/build.sh:
```diff
@@ -0,0 +1,36 @@
+mkdir -p "${BUILD_DIR}"
+cd "${BUILD_DIR}"
+
+if [[ -f configure-done ]]; then
+    echo "Skipping configure step, already done."
```
This is to enable incremental compilation in the simplest way: don't configure if we have already configured in that build directory. This seems sufficient, as nuking the build dir may be a good idea anyway when reconfiguring. But there might be a different / more flexible way to do this, not sure.

FFY00 replied:
`make` should re-run `configure` already if needed, so this approach seems fine. Alternatively, we could run `configure` unconditionally with `--config-cache`, which should make it much faster, but I prefer the current option.

lucascolley commented on Tools/pixi-packages/build.sh:
```diff
@@ -0,0 +1,36 @@
+    echo "BUILD TYPE: DEFAULT"
+    BUILD_DIR="../build"
+    CONFIGURE_EXTRA=""
+fi
```
right now we switch cases on `PYTHON_ASAN`, but if more variants are added we may want to introduce a new variable specifically for this variant selection.

ngoldbaum replied:
IMO it's worth doing this for TSan and maybe UBSan as well, so maybe better to define it terms of the kind of sanitizer being used. Maybe call it `PYTHON_SANITIZER`? ASan and TSan are mutually exclusive but ASan and UBsan can be used at the same time to make it extra annoying...

lucascolley replied:
maybe there could be other useful variants aside from sanitizers too?

FFY00 replied:
Yes, though IMO it'd be better to scope this PR for ASAN only, and then add a follow-up adding other kinds.

ngoldbaum replied:
Fair enough, I'll do TSan in a followup given this advice.

Maybe for now call it `PYTHON_VARIANT` and only implement the "asan" case?

lucascolley replied:
switched to `PYTHON_VARIANT` in https://github.com/python/cpython/pull/142469/commits/ad7a35f619b648286a698d0042213ba03d13d439

lucascolley commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,67 @@
+context:
+  # XXX: keep up to date
+  version: "3.15"
+  minor_version: "3.15"
```
should be kept up to date with https://github.com/python/cpython/blob/bcb6cbdf3bec6a856b739587117963739b80156e/configure#L607

`minor_version` is so that we can

```bash
ln -sf "${PREFIX}/bin/python${MINOR_VERSION}" "${PREFIX}/bin/python"
```

in the build script

FFY00 replied:
Isn't there any way to determine this dynamically? Via the source, or git, for eg.

> should be kept up to date with
> 
> [cpython/configure](https://github.com/python/cpython/blob/bcb6cbdf3bec6a856b739587117963739b80156e/configure#L607)
> 
> Line 607 in [bcb6cbd](/python/cpython/commit/bcb6cbdf3bec6a856b739587117963739b80156e)
>  PACKAGE_VERSION='3.15' 
> 
> `minor_version` is so that we can
> 
> ```
> ln -sf "${PREFIX}/bin/python${MINOR_VERSION}" "${PREFIX}/bin/python"
> ```
> 
> in the build script

If we move the symlink target to `python3`, we can drop the need for `MINOR_VERSION`.

lucascolley replied:
> we move the symlink target to `python3`, we can drop the need for `MINOR_VERSION`.

ah yea, if `python3` always exists then perfect

FFY00 replied:
As long as you use `make install`, instead of `make altinstall`, it does :blush:

lucascolley replied:
remove `MINOR_VERSION` in https://github.com/python/cpython/pull/142469/commits/ad7a35f619b648286a698d0042213ba03d13d439.

> Isn't there any way to determine this dynamically? Via the source, or git, for eg.

maybe, but I'm not sure how

FFY00 replied:
We can leave `version` in for now, but let's ping the pixi team to see if it can be set dynamically.

lucascolley replied:
I have asked offline

lucascolley replied:
We should be able to use https://rattler-build.prefix.dev/latest/experimental_features/#load_from_filefile_path

lucascolley commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,67 @@
+
+# XXX: maybe not all requirements necessary
+# derived from https://github.com/conda-forge/python-feedstock/blob/main/recipe/meta.yaml
+requirements:
```
maintenance may be needed here when dependencies change.

Also, not sure if there are other optional dependencies that we would want to include?

StanFromIreland commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,2 @@
+- see comments for which fields should be maintained and kept up-to-date
```
When I read this, I still have no idea what the stuff in this folder is doing/for, can you please add a general overview. Also, you need to add an entry to the top-level `Tools/README`

lucascolley replied:
yeah of course, just opened this for preliminary discussion first.

lucascolley replied:
@StanFromIreland please could you review https://github.com/python/cpython/pull/142469/commits/b49095c1bc6ffdd7246e5aef4b99ab9e6411cf3b ? Thanks!

lucascolley committed:
Merge branch 'main' into pixi-build

lucascolley commented on Tools/pixi-packages/build.sh:
```diff
@@ -0,0 +1,36 @@
+    echo "BUILD TYPE: ASAN"
+    BUILD_DIR="../build_asan"
+    CONFIGURE_EXTRA="--with-address-sanitizer"
+    export ASAN_OPTIONS="detect_leaks=0:symbolize=1:strict_init_order=true:allocator_may_return_null=1:use_sigaltstack=0"
```
per https://github.com/scipy/scipy/pull/24066#discussion_r2604118318
```suggestion
    export ASAN_OPTIONS="strict_init_order=true"
```

lucascolley committed:
tidy gitignore

lucascolley committed:
respond to review
- switch cases on `PYTHON_VARIANT`
- remove `minor_version` by using `python3`
- remove runtime-only asan options

lucascolley committed:
README updates

lucascolley committed:
use `.md` to preview rendering

StanFromIreland commented on Tools/pixi-packages/README:
```diff
@@ -0,0 +1,36 @@
+## CPython Pixi Packages
```
If this is to be markdown please add the markdown file extension so it automatically renders when possible (e.g. GitHub's viewer).

lucascolley replied:
https://github.com/python/cpython/pull/142469/commits/4d0983ef3e49d205d82558311586b89f9d6b7ca9 :)

hugovk commented on Tools/README:
```diff
@@ -43,6 +43,8 @@ patchcheck      Tools for checking and applying patches to the Python source cod
 
 peg_generator   PEG-based parser generator (pegen) used for new parser.
 
+pixi-packages   Pixi package definitions for downstream from-source builds
```
```suggestion
pixi-packages   Pixi package definitions for downstream from-source builds.
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+This directory contains definitions for [Pixi packages](https://pixi.sh/latest/reference/pixi_manifest/#the-package-section)
+which can be built from the CPython source code.
+
+Downstream developers can make use of these packages by adding them as git dependencies in a
```
```suggestion
Downstream developers can make use of these packages by adding them as Git dependencies in a
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+
+This is particularly useful when developers need to build CPython from source
+(for example, for an ASan-instrumented build), as it does not require any manual
+cloning or building steps. Instead, Pixi will automatically handle both the building
```
```suggestion
clone or build steps. Instead, Pixi will automatically handle both the build
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+
+- the `version` fields in each `recipe.yaml` should be kept up to date with the Python version
+- dependency requirements should be kept up to date in each `recipe.yaml`
+- `build.sh` should be updated for any breaking changes in the `configure` and `make` workflow
```
```suggestion
- Keep the `version` fields in each `recipe.yaml` up to date with the Python version
- Keep the dependency requirements up to date in each `recipe.yaml`
- Update `build.sh` for any breaking changes in the `configure` and `make` workflow
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+- more package variants (e.g. TSan, UBSan)
+- support for Windows
+- using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on https://github.com/prefix-dev/pixi/issues/4599
+- a workaround can be removed from the build script once https://github.com/prefix-dev/rattler-build/issues/2012 is resolved
```
```suggestion
- More package variants (such as TSan, UBSan)
- Support for Windows
- Using a single `pixi.toml` and `recipe.yaml` for all package variants is blocked on https://github.com/prefix-dev/pixi/issues/4599
- A workaround can be removed from the build script once https://github.com/prefix-dev/rattler-build/issues/2012 is resolved
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+## CPython Pixi Packages
```
```suggestion
# CPython Pixi packages
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+- `default`
+- `asan`: ASan-instrumented build with `PYTHON_ASAN=1`
+
+### Maintenance
```
```suggestion
## Maintenance
```

hugovk commented on Tools/pixi-packages/README.md:
```diff
@@ -0,0 +1,36 @@
+- dependency requirements should be kept up to date in each `recipe.yaml`
+- `build.sh` should be updated for any breaking changes in the `configure` and `make` workflow
+
+### Opportunities for future improvement
```
```suggestion
## Opportunities for future improvement
```

hugovk commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,65 @@
+    env:
+      PYTHON_VARIANT: "asan"
+
+# XXX: maybe not all requirements necessary
```
Can we check and only include the necessary ones?

FFY00 replied:
I don't think `libtool` is needed on POSIX, but that should be all.

FFY00 replied:
I confirm `libtool` isn't needed.

hugovk commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,65 @@
+context:
+  # XXX: keep up to date
```
Re: https://discuss.python.org/t/cryptic-sequences-in-comments/38694

Can we replace all the "XXX" with "TODO" or something else? Maybe we don't even need it here:

```suggestion
  # Keep up to date
```

If we do merge this, we have code in https://github.com/python/release-tools which is run during the first beta, when `main` becomes the next version, that updates a bunch of versions in the repo. And we can later update that to also update these versions.

lucascolley replied:
https://rattler-build.prefix.dev/latest/experimental_features/#load_from_filefile_path may remove the need for this

lucascolley replied:
Is there a YAML, TOML, or JSON file that contains the version @hugovk ? If not, could we add one?

hugovk replied:
There isn't. Yes we can add one, it removes duplication here, and will be easier to programmatically update such a file.

lucascolley replied:
any preference for format / location / name?

hugovk replied:
Sorry, I missed this because it was marked as resolved but I don't think it has been :)

Format: TOML or JSON, we can read using the stdlib.

Location: `Tools/pixi-packages/`, or `Tools/` so other scripts can re-use it.

Name: Hmm, `python-version.[ext]` ?

lucascolley replied:
I think it was resolved for this PR via the third paragraph of https://github.com/python/cpython/pull/142469#pullrequestreview-3563946286

hugovk replied:
Okay, thanks!

hugovk commented on Tools/pixi-packages/default/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+context:
+  # XXX: keep up to date
```
```suggestion
  # Keep up to date
```

hugovk commented on Tools/pixi-packages/default/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+  script:
+    file: ../build.sh
+
+# XXX: maybe not all requirements necessary
```
Let's check and only include the necessary ones.

lucascolley committed:
Apply suggestions from code review

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

FFY00 commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,65 @@
+    - pkg-config
+    # configure script looks for llvm-ar for lto
+    - if: osx
+      then:
+        - llvm-tools
```
Do we still need this here if we ask for `llvm-tools-19` below?

lucascolley replied:
maybe. The comment above suggests this is here for `llvm-ar`. [`llvm-tools-19` contains `llvm-ar-19` but not `llvm-ar`](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fllvm-tools-19-19.1.7-h3b15d91_2.conda#files-77), while [`llvm-tools` contains `llvm-ar`](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fllvm-tools-21.1.7-hb700be7_0.conda#files-77).

FFY00 replied:
Let's keep it in then.

FFY00 commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,65 @@
+    - ${{ compiler('c') }}
+    - ${{ compiler('cxx') }}
+    - make
+    - libtool
```
```suggestion
```

FFY00 commented on Tools/pixi-packages/default/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+    - ${{ compiler('c') }}
+    - ${{ compiler('cxx') }}
+    - make
+    - libtool
```
```suggestion
```

FFY00 commented on Tools/pixi-packages/default/recipe.yaml:
```diff
@@ -0,0 +1,63 @@
+  script:
+    file: ../build.sh
+
+# XXX: maybe not all requirements necessary
```
```suggestion
```

lucascolley committed:
Apply suggestion from @FFY00

Co-authored-by: Filipe Laíns <filipe.lains@gmail.com>

lucascolley committed:
Apply suggestion from @FFY00

Co-authored-by: Filipe Laíns <filipe.lains@gmail.com>

lucascolley committed:
Apply suggestion from @FFY00

Co-authored-by: Filipe Laíns <filipe.lains@gmail.com>

lucascolley commented on Tools/pixi-packages/asan/recipe.yaml:
```diff
@@ -0,0 +1,64 @@
+    env:
+      PYTHON_VARIANT: "asan"
+
+# XXX: maybe not all requirements necessary
```
```suggestion
```

lucascolley committed:
Apply suggestion from @lucascolley

FFY00 reviewed (approved):
Thanks so much for working on this @lucascolley! This is PR looks great, and sets up to add other sanitizer variants.

The `pixi` limitation that requires us to have separate recipes for each variant (https://github.com/prefix-dev/pixi/issues/4599) is unfortunate, but I think doable from a maintenance POV, given that we have a few people interested in this. We'll see how it translates in practice, and hope the feature request gets implemented in the meantime.

I opened https://github.com/prefix-dev/rattler-build/issues/2047 regarding the version string that currently needs to be updated manually, but could be dirtermined from git, with additional information. But let's wait for the `pixi` [jinja functions](https://rattler-build.prefix.dev/latest/experimental_features/#jinja-functions) to get stabilized before adding them to our recipes.

I'll leave this PR open a couple days to give a bit more time for folks to give feedback, but otherwise, this looks good to go 😊
